engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/NadelServiceTypeFilterTransform.kt

```diff
@@ -0,0 +1,144 @@
+import graphql.nadel.enginekt.transform.artificial.NadelAliasHelper
+import graphql.nadel.enginekt.transform.query.NadelQueryPath
+import graphql.nadel.enginekt.transform.query.NadelQueryTransformer
+import graphql.nadel.enginekt.transform.result.NadelResultInstruction
+import graphql.nadel.enginekt.transform.result.json.JsonNodeExtractor
+import graphql.nadel.enginekt.util.AnyMap
+import graphql.nadel.enginekt.util.JsonMap
+import graphql.nadel.enginekt.util.queryPath
+import graphql.nadel.enginekt.util.resolveObjectTypes
+import graphql.nadel.enginekt.util.toBuilder
+import graphql.normalized.ExecutableNormalizedField
+import graphql.normalized.ExecutableNormalizedField.newNormalizedField
+
+class NadelServiceTypeFilterTransform : NadelTransform<State> {
+    data class State(
+        val aliasHelper: NadelAliasHelper,
+        val typeNamesOwnedByService: Set<String>,
+        val fieldObjectTypesOwnedByService: List<String>,
+        val overallField: ExecutableNormalizedField,
+    )
+
+    override suspend fun isApplicable(
+        executionContext: NadelExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        services: Map<String, Service>,
+        service: Service,
+        overallField: ExecutableNormalizedField,
+    ): State? {
+        when {
+            // Ignore top level fields, they won't belong to multiple services
+            // Do not randomly remove this, we rely on this later on too.
+            // This transform never runs for top level fields
+            overallField.parent == null -> return null
+            // Introspection - do nothing. INTROSPECTION REALLY SHOULDN'T BE EXECUTING THIS CODE PATH
+            service.name == IntrospectionService.name -> return null
+        }
+
+        val typeNamesOwnedByService = getOwnedTypeNames(executionBlueprint, service)
+        val typesOwnedByService = overallField.objectTypeNames
+            .filter { it in typeNamesOwnedByService }
+
+        // All types are owned by service
+        if (typesOwnedByService.size == overallField.objectTypeNames.size) {
+            return null
+        }
+
+        return State(
+            aliasHelper = NadelAliasHelper.forField(
+                tag = "type_filter",
+                field = overallField,
+            ),
+            typeNamesOwnedByService = typeNamesOwnedByService,
+            fieldObjectTypesOwnedByService = typesOwnedByService,
+            overallField = overallField,
+        )
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        transformer: NadelQueryTransformer.Continuation,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+        // Nothing to query if there are no fields, we need to add selection
+        if (state.fieldObjectTypesOwnedByService.isEmpty()) {
```

Comment:
So, in case all types have been filtered out we add one artificial field per object type associated with the original field? 
And those artificial fields will later be discarded from the result data?
What's the purpose of this?

Reply:
Yea it's the same issue we have with removed fields. Basically take this query

```graphql
query {
    test {
        ... on Test {
            id
        }
        ... on QueryError {
            extensions {
                ... on LabQueryErrorExtension {
                    statusCode
                }
            }
        }
    }
}
```

There are two services `Test` and `Lab`. The `LabQueryErrorExtension` gets removed because it doesn't belong to the `Test` service which the query is going to.

In that case we end up with this query

```graphql
query {
    test {
        ... on Test {
            id
        }
        ... on QueryError {
            extensions
        }
    }
}
```

Which isn't legal because `extensions` needs subselections. So in this case we add `__typename` as a placeholder so that we still execute the `extensions` field.

Reply:
I see. thanks for the explanation!